subsys/usb/usb_c/usbc_pe_common.c

```diff
@@ -0,0 +1,848 @@
+	 *   * Protocol Errors occurred on SOP during an Interruptible AMS or
+	 *   * Protocol Errors occurred on SOP during any AMS where the first
+	 *     Message in the sequence has not yet been sent i.e. an unexpected
+	 *     Message is received instead of the expected GoodCRC Message
+	 *     response.
+	 */
+	/* All error types besides transmit errors are Protocol Errors. */
+	if ((e != ERR_XMIT &&
+	     atomic_test_bit(&pe->flags, PE_FLAGS_INTERRUPTIBLE_AMS) == false) ||
+	     e == ERR_XMIT ||
+	     (atomic_test_bit(&pe->flags, PE_FLAGS_EXPLICIT_CONTRACT) == false &&
+	     type == PD_PACKET_SOP)) {
+		policy_notify(dev, PROTOCOL_ERROR);
+		pe_send_soft_reset(dev, type);
+	}
+	/*
+	 * Transition to PE_Snk_Ready by a Protocol
+	 * Error during an Interruptible AMS.
+	 */
+	else {
+		pe_set_state(dev, PE_SNK_READY);
+	}
+}
+
+/**
+ * @brief Informs the Policy Engine of a discard.
+ */
+void pe_report_discard(const struct device *dev)
+{
+	struct usbc_port_data *data = dev->data;
+	struct policy_engine *pe = data->pe;
+
+	/*
+	 * Clear local AMS indicator as our AMS message was discarded, and flag
+	 * the discard for the PE
+	 */
+	atomic_clear_bit(&pe->flags, PE_FLAGS_DPM_INITIATED_AMS);
+	atomic_set_bit(&pe->flags, PE_FLAGS_MSG_DISCARDED);
+}
+
+/**
+ * @brief Called by the Protocol Layer to informs the Policy Engine
+ *	  that a message has been received.
+ */
+void pe_message_received(const struct device *dev)
+{
+	struct usbc_port_data *data = dev->data;
+	struct policy_engine *pe = data->pe;
+
+	atomic_set_bit(&pe->flags, PE_FLAGS_MSG_RECEIVED);
+}
+
+/**
+ * @brief Informs the Policy Engine that a hard reset was received.
+ */
+void pe_got_hard_reset(const struct device *dev)
+{
+	pe_set_state(dev, PE_SNK_TRANSITION_TO_DEFAULT);
+}
+
+/**
+ * @brief Informs the Policy Engine that a hard reset was sent.
+ */
+void pe_hard_reset_sent(const struct device *dev)
+{
+	struct usbc_port_data *data = dev->data;
+	struct policy_engine *pe = data->pe;
+
+	atomic_clear_bit(&pe->flags, PE_FLAGS_HARD_RESET_PENDING);
+}
+
+/**
+ * @brief Indicates if an explicit contract is in place
+ */
+bool pe_is_explicit_contract(const struct device *dev)
+{
+	struct usbc_port_data *data = dev->data;
+	struct policy_engine *pe = data->pe;
+
+	return atomic_test_bit(&pe->flags, PE_FLAGS_EXPLICIT_CONTRACT);
+}
+
+/**
+ * @brief Return true if the PE is is within an atomic messaging sequence
+ *	  that it initiated with a SOP* port partner.
+ */
+bool pe_dpm_initiated_ams(const struct device *dev)
+{
+	struct usbc_port_data *data = dev->data;
+	struct policy_engine *pe = data->pe;
+
+	return atomic_test_bit(&pe->flags, PE_FLAGS_DPM_INITIATED_AMS);
+}
+
+/** Private Policy Engine Layer API below */
+
+/**
+ * @brief Sets a Policy Engine state
+ */
+void pe_set_state(const struct device *dev,
+			 const enum usbc_pe_state state)
+{
+	struct usbc_port_data *data = dev->data;
+
+	smf_set_state(SMF_CTX(data->pe), &pe_states[state]);
```

Comment:
This need a bounds check that `state < ARRAY_SIZE(pe_states)`.  You could also wrap this in a macro to do a build time assert (see the `tc_set_state()` example).

Reply:
Same error with ARRAY_SIZE(pe_states)

Reply:
Is the build error from the run-time check or the build-time check?  `__ASSERT(state < ARRAY_SIZE(pe_states), "invalid pe_state %d", state);` should work.

Reply:
Yes that worked. I was using BUILD_ASSERT